smallrye-reactive-messaging-provider/src/main/java/io/smallrye/reactive/messaging/extension/EmitterImpl.java

```diff
@@ -67,18 +73,18 @@ boolean isConnected() {
     }
 
     @Override
-    public synchronized Emitter<T> send(T msg) {
+    public synchronized CompletionStage<Void> send(T msg) {
         if (msg == null) {
             throw new IllegalArgumentException("`null` is not a valid value");
         }
         FlowableEmitter<Message<? extends T>> emitter = verify();
         if (msg instanceof Message) {
             //noinspection unchecked
-            emitter.onNext((Message) msg);
+            return CompletableFuture.runAsync(() -> emitter.onNext((Message) msg));
```

Comment:
That's not correct.

You need something like (be aware this is pseudo-code):

```
CompletableFuture future = new CompletableFuture();
if (msg instanceof Message) {
    // here things need to be discussed
    // if the user passes a Message, he can attach the acknowledgment logic
   // however we cannot really do anything. We would need a way to customize the Supplier
} else {
   emitter.onNext(Message.of(msg, () -> future.complete(null));
}
```

Reply:
updated